src/Core/Datatypes/DyadicTensor.h

```diff
@@ -0,0 +1,322 @@
+   and/or sell copies of the Software, and to permit persons to whom the
+   Software is furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef CORE_DATATYPES_DYADIC_TENSOR_H
+#define CORE_DATATYPES_DYADIC_TENSOR_H
+
+#include <Core/Datatypes/TensorBase.h>
+#include <Core/Datatypes/TensorFwd.h>
+#include <Core/Utils/Exception.h>
+#include <Eigen/Dense>
+#include <algorithm>
+#include <initializer_list>
+#include <limits>
+#include <Core/Datatypes/share.h>
+
+namespace SCIRun {
+namespace Core {
+  namespace Datatypes {
+    // Dyadic tensors are also known as second-order tensors
+    template <typename Number, size_t Dim>
+    class DyadicTensorGeneric : public TensorBaseGeneric<Number, Eigen::Sizes<Dim, Dim>>
+    {
+     public:
+      using parent = TensorBaseGeneric<Number, Eigen::Sizes<Dim, Dim>>;
+      using VectorType = Eigen::Matrix<Number, Dim, 1>;
+      using MatrixType = Eigen::Matrix<Number, Dim, Dim>;
+
+      DyadicTensorGeneric() : parent() { parent::setZero(); }
+
+      explicit DyadicTensorGeneric(Number val) : parent()
+      {
+        for (size_t i = 0; i < Dim; ++i)
+          for (size_t j = 0; j < Dim; ++j)
+            (*this)(i, j) = (i == j) ? val : 0;
+      }
+
+      explicit DyadicTensorGeneric(const std::vector<VectorType>& eigvecs) : parent()
+      {
+        if (eigvecs.size() != Dim)
+          THROW_INVALID_ARGUMENT("The number of input vectors must be " + Dim);
+        setEigenVectors(eigvecs);
+      }
+
+      DyadicTensorGeneric(const std::initializer_list<VectorType>& eigvecs) : parent()
+      {
+        if (eigvecs.size() != Dim)
+          THROW_INVALID_ARGUMENT("The number of input vectors must be " + Dim);
+        setEigenVectors(eigvecs);
+      }
+
+      DyadicTensorGeneric(const DyadicTensorGeneric<Number, Dim>& other) : parent()
+      {
+        for (size_t i = 0; i < Dim; ++i)
+          for (size_t j = 0; j < Dim; ++j)
+            (*this)(index(i), index(j)) = other(index(i), index(j));
+        eigvecs_ = other.getEigenvectors();
+        eigvals_ = other.getEigenvalues();
+        haveEigens_ = true;
+      }
+
+      DyadicTensorGeneric(DyadicTensorGeneric<Number, Dim>&& other) : parent()
+      {
+        auto otherData = other.data();
+        std::move(otherData, otherData + other.size(), this->data());
```

Comment:
Here's the new move statement. I confirmed this was being called from the new test I added.

Reply:
That'll still just do a copy over the range for doubles since they're not movable one-by-one, but that's OK; if you ever make a DyadicTensorGeneric<std::thread>, for instance, this constructor will be most useful 🥇

Reply:
Oh so it just calls move over each item in the iteration, essentially copying the doubles?

Reply:
Yeah it's just a range move, just like std::copy is a range copy--so it will call move on each element from begin to end. Good for large movable objects in a container, but still not moving a whole block at a time sort of thing (like std::vector's move ctor)

Reply:
It's still an improvement, so let's just get it merged and optimize that later

Reply:
Sounds good. At the very least, I can add this range copy to the copy constructor.